snapcraft/internal/deltas/_deltas.py

```diff
@@ -0,0 +1,199 @@
+
+    This class is responsible for the snap delta file generation
+    """
+
+    delta_format = None
+    delta_file_extname = 'delta'
+    delta_tool_path = None
+
+    def __init__(self, source_path, target_path):
+        self.source_path = source_path
+        self.target_path = target_path
+        self.pre_check()
+
+    def pre_check(self):
+        self._check_properties()
+        self._check_file_existence()
+        self._check_delta_gen_tool()
+
+    def _check_properties(self):
+        if self.delta_format is None:
+            raise DeltaFormatIsNoneError(
+                'delta_format must be set in subclass')
+        if self.delta_tool_path is None:
+            raise DeltaToolPathIsNoneError(
+                'delta_too_path must be set in subclass')
+        if self.delta_format not in delta_formats_options:
+            raise DeltaFormatOptionError(
+                'delta_format must be option in {}'.format(
+                    delta_formats_options))
+
+    def _check_file_existence(self):
+        if not os.path.exists(self.source_path):
+            raise ValueError(
+                'source file {} not exist'.format(self.source_path))
+
+        if not os.path.exists(self.target_path):
+            raise ValueError(
+                'target file {} not exist'.format(self.target_path))
+
+    def _check_delta_gen_tool(self):
+        """check if the delta generation tool exists"""
+        if not executable_exists(self.delta_tool_path):
+            delta_path = shutil.which(self.delta_format)
+            if not delta_path:
+                raise DeltaGenerationError(
+                    "Could not find {} executable.".format(
+                        self.delta_format))
+            else:
+                self.delta_tool_path = delta_path
+
+    def find_unique_file_name(self, path_hint):
+        """Return a path on disk similar to 'path_hint' that does not exist.
+
+        This function can be used to ensure that 'path_hint' points to a file
+        on disk that does not exist. The returned filename may be a modified
+        version of 'path_hint' if 'path_hint' already exists.
+
+        """
+        target = path_hint
+        counter = 0
+        while os.path.exists(target):
+            target = "{}-{}".format(path_hint, counter)
+            counter += 1
+        return target
+
+    def _setup_std_output(self, source_path):
+        """helper to setup the stdout and stderr for subprocess"""
+        workdir = os.path.dirname(source_path)
+
+        stdout_path = self.find_unique_file_name(
+            os.path.join(workdir, '{}-out'.format(self.delta_format)))
+        stdout_file = open(stdout_path, 'wb')
+
+        stderr_path = self.find_unique_file_name(
+            os.path.join(workdir, '{}-err'.format(self.delta_format)))
+        stderr_file = open(stderr_path, 'wb')
+
+        return workdir, stdout_path, stdout_file, stderr_path, stderr_file
+
+    def make_delta(self, output_dir=None):
```

Comment:
this needs a parameter to pass in an indicator (which the caller can pass in) as I bet this will take some time.